kyo-reactive-streams/shared/src/main/scala/kyo/interop/reactive-streams/StreamSubscriber.scala

```diff
@@ -0,0 +1,283 @@
+                        if state.compareAndSet(curState, nextState) then
+                            IO(Result.success(leftOver))
+                        else
+                            handlePoll
+                        end if
+                case other =>
+                    if state.compareAndSet(curState, other) then
+                        IO(Result.success(Chunk.empty))
+                    else
+                        handlePoll
+            end match
+        end handlePoll
+        IO(handlePoll)
+    end poll
+
+    private[interop] def interupt(using Frame): Unit < IO =
+        @tailrec def handleInterupt: Unit < IO =
+            val curState = state.get()
+            curState match
+                case (UpstreamState.Uninitialized, maybePromise) =>
+                    val nextState = UpstreamState.Finished(Absent, Chunk.empty) -> Absent
+                    if state.compareAndSet(curState, nextState) then
+                        IO(maybePromise.foreach(_.completeDiscard(Result.success(()))))
+                    else
+                        handleInterupt
+                    end if
+                case (UpstreamState.WaitForRequest(subscription, _, _), Absent) =>
+                    val nextState = UpstreamState.Finished(Absent, Chunk.empty) -> Absent
+                    if state.compareAndSet(curState, nextState) then
+                        IO(subscription.cancel())
+                    else
+                        handleInterupt
+                    end if
+                case other =>
+                    if state.compareAndSet(curState, other) then
+                        IO.unit
+                    else
+                        handleInterupt
+            end match
+        end handleInterupt
+        IO(handleInterupt)
+    end interupt
+
+    private[interop] def emit(using Frame, Tag[Emit[Chunk[V]]]): Ack < (Emit[Chunk[V]] & Async) =
+        Emit.andMap(Chunk.empty) { ack =>
+            Loop(ack) {
+                case Ack.Stop => interupt.andThen(Loop.done(Ack.Stop))
+                case Ack.Continue(_) =>
+                    await
+                        .map {
+                            case true => request.andThen(Ack.Continue())
+                            case false => poll.map {
+                                    case Result.Success(nextChunk)  => Emit(nextChunk)
+                                    case Result.Error(e: Throwable) => Abort.panic(e)
+                                    case _                          => Ack.Stop
+                                }
+                        }
+                        .map(Loop.continue(_))
+            }
+        }
+
+    def stream(using Frame, Tag[Emit[Chunk[V]]]): Stream[V, Async] = Stream(emit)
+
+end StreamSubscriber
+
+object StreamSubscriber:
+
+    abstract private[reactivestreams] class SubscriberDone
+    private[reactivestreams] case object SubscriberDone extends SubscriberDone
+
+    enum EmitStrategy derives CanEqual:
```

Comment:
introduce `EmitStrategy` so that users can choose the behavior they want

Reply:
Nice! We could eventually introduce more advanced strategies including timeouts for example